### PR TITLE
feat: Add file resource blacklist [DHIS2-9157]

### PR DESCRIPTION
**Pull request by larshelge**
Introduces class `FileResourceBlocklist`.

Disallowed content types:

| Content type | Content type |
| ------------------------------------- | --- |
| text/html                             | application/x-ms-dos-executable |
| text/css                              | application/vnd.microsoft.portable-executable |
| text/javascript                       | application/vnd.apple.installer+xml |
| font/otf                              | application/vnd.mozilla.xul+xml |
| application/x-shockwave-flash         | application/x-httpd-php  |
| application/vnd.debian.binary-package | application/x-sh |
| application/x-rpm                     | application/x-csh |
| application/java-archive              |  |

Disallowed file extensions:

| File extension | File extension | File extension |
| ---- | ---- | ---- |
| html | deb  | xul  |
| htm  | rpm  | php  |
| css  | jar  | bin  |
| js   | jsp  | sh   |
| mjs  | exe  | csh  |
| otf  | msi  |  bat  |
| swf  | mpkg |      |

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        "jar",
+        "jsp",
+        "exe",
+        "msi",
+        "mpkg",
+        "xul",
+        "php",
+        "bin",
+        "sh",
+        "csh"
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:81]: Maybe, add "com" and "bat" files as well?

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        FileResource frA = new FileResource( "Click_Me.exe", "application/x-ms-dos-executable", 451, "", FileResourceDomain.DATA_VALUE );
+        FileResource frB = new FileResource( "evil_script.sh", "application/pdf", 125, "", FileResourceDomain.MESSAGE_ATTACHMENT ); // Fake content type
+        FileResource frC = new FileResource( "cookie_stealer", "text/javascript", 631, "", FileResourceDomain.USER_AVATAR ); // No file extension
+
+        assertFalse( FileResourceBlacklist.isValid( frA ) );
+        assertFalse( FileResourceBlacklist.isValid( frB ) );
+        assertFalse( FileResourceBlacklist.isValid( frC ) );
+    }
+}
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java:64]: Should we add a test to cover the cases where the client passes in a null argument to the "isValid" method?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.fileresource;
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:29]: Usually package declaration is before the Copyright para.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.fileresource;
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java:29]: package declaration before copyright?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/DefaultFileResourceService.java [hunk 5]
```diff
@@ -70,12 +66,6 @@
 
     public static final Predicate<FileResource> IS_ORPHAN_PREDICATE = (fr -> !fr.isAssigned());
 
-    private static final Set<String> CONTENT_TYPE_BLACKLIST = Sets.newHashSet( "text/html",
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/DefaultFileResourceService.java:73]: Should we also move away from naming "Blacklist", and using "Blocklist" instead? 😄 
Read a lot of recent changes proposed in several companies/projects that want to move away from using "Black" as bad and "White" as good.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 3 above)
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:29]: package before license

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+import org.apache.commons.io.FilenameUtils;
+
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:32]: extra newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Indicates whether the given file resource has a valid file extension and content type
+     * according to the blacklist.
+     *
+     * @param fileResource the {@link FileResource}.
+     * @return true if valid, false if invalid.
+     */
+    public static boolean isValid( FileResource fileResource )
+    {
```
Comment (netroms) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:92]: Add checkNotNull(fileResource)?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:29]: I have sort of changed my mind on this one. Seems the Java convention is to have copyright before package declaration. It also works better with auto-organize in the IDE (the opposite does not work properly). I think we should do a switch across the code base at some point @ameenhere .

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 1 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:81]: Yes agree, will add at least bat.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java (hunk 2 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/fileresource/FileResourceBlacklistTest.java:64]: Yes agree, will fix.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 6 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:29]: I have sort of changed my mind on this one. Seems the Java convention is to have copyright before package declaration. It also works better with auto-organize in the IDE (the opposite does not work properly). I think we should do a switch across the code base at some point.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java (hunk 7 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/FileResourceBlacklist.java:92]: Agree, included null check for FileResource.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/DefaultFileResourceService.java (hunk 5 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/fileresource/DefaultFileResourceService.java:73]: Sure, changed to Blocklist.
